File: analyzers/tests/SonarAnalyzer.Test/TestCases/ExpectedExceptionAttributeShouldNotBeUsed.MsTest.vb
Imports System
Imports Microsoft.VisualStudio.TestTools.UnitTesting

Public Class ExceptionTests

    ' Noncompliant@+2
    <TestMethod>
    <ExpectedException(GetType(DivideByZeroException))>
    Public Sub ExpectedExceptionAttrbutesShouldNotBeUsedOnSub()
        Dim instance As Integer = 42
        instance.ToString()
    End Sub

    <TestMethod>
    <ExpectedExceptionAttribute(GetType(ArgumentNullException))>  ' Noncompliant
    Public Sub WithAttributeSuffix()
        Dim x As Boolean = True
        x.ToString()
    End Sub

    <TestMethod>
    <Microsoft.VisualStudio.TestTools.UnitTesting.ExpectedException(GetType(ArgumentNullException))>  ' Noncompliant
    Public Sub FullyQualifiedAttribute()
        Dim x As Boolean = True
        x.ToString()
    End Sub

    <TestMethod>
    <Unrelated.ExpectedException(GetType(ArgumentNullException))>  ' Noncompliant - FPgit
    Public Sub UnrelatedAttribute()
        Dim x As Boolean = True
        x.ToString()
    End Sub

    ' Noncompliant@+2
    <TestMethod>
    <ExpectedException(GetType(DivideByZeroException))>
    Public Function ExpectedExceptionAttrbutesShouldNotBeUsedOnFunction() As String
        Dim instance As Integer = 42
        Return instance.ToString()
    End Function

    ' Compliant - one line
    <TestMethod>
    <ExpectedException(GetType(DivideByZeroException))>
    Public Sub ExpectedExceptionAttrbuteAllowedForSingleLineSub()
        Throw New NotImplementedException()
    End Sub

    ' Compliant - one line
    <TestMethod>
    <ExpectedException(GetType(DivideByZeroException))>
    Public Function ExpectedExceptionAttrbuteAllowedForSingleLineFunction() As Integer
        Return 42
    End Function
End Class

' https://github.com/SonarSource/sonar-dotnet/issues/8300
Class Repro_8300
    <TestMethod>
    <ExpectedException(GetType(InvalidOperationException))> ' Compliant - using ExpectedException makes the test more readable
    Public Sub AssertInFinally()
        Console.ForegroundColor = ConsoleColor.Red
        Try
            Throw New InvalidOperationException()
        Finally
            Assert.AreEqual(ConsoleColor.Black, Console.ForegroundColor)
        End Try
    End Sub

    <TestMethod>
    <ExpectedException(GetType(InvalidOperationException))> ' Noncompliant
    Public Sub NoAssertInFinally()
        Console.ForegroundColor = ConsoleColor.Red
        Try
            Throw New InvalidOperationException()
        Finally
            Console.WriteLine("No Assert")
        End Try
    End Sub

    <TestMethod>
    <ExpectedException(GetType(InvalidOperationException))> ' Noncompliant
    Public Sub NoAssertInCatch()
        Console.ForegroundColor = ConsoleColor.Red
        Try
            Throw New InvalidOperationException()
        Catch e As InvalidOperationException
            Console.ForegroundColor = ConsoleColor.Black
        End Try
    End Sub

    <TestMethod>
    <ExpectedException(GetType(InvalidOperationException))> ' Compliant
    Public Sub AssertInCatch()
        Console.ForegroundColor = ConsoleColor.Red
        Try
            Throw New InvalidOperationException()
        Catch e As InvalidOperationException
            Assert.AreEqual(ConsoleColor.Black, Console.ForegroundColor)
        End Try
    End Sub

    <TestMethod>
    <ExpectedException(GetType(InvalidOperationException))> ' Compliant
    Public Sub AssertInAllCatch()
        Console.ForegroundColor = ConsoleColor.Red
        Try
            Throw New InvalidOperationException()
        Catch
            Assert.AreEqual(ConsoleColor.Black, Console.ForegroundColor)
        End Try
    End Sub

    <TestMethod>
    <ExpectedException(GetType(InvalidOperationException))> ' Compliant
    Public Sub AssertInAllCatch_InvocationBeforeAssert()
        Console.ForegroundColor = ConsoleColor.Red
        Try
            Throw New InvalidOperationException()
        Catch
            Console.WriteLine("An invocation before Assert")
            Assert.AreEqual(ConsoleColor.Black, Console.ForegroundColor)
        End Try
    End Sub

    <TestMethod>
    <ExpectedException(GetType(InvalidOperationException))> ' Compliant
    Public Sub AssertInAllCatch_InvocationAfterAssert()
        Console.ForegroundColor = ConsoleColor.Red
        Try
            Throw New InvalidOperationException()
        Catch
            Assert.AreEqual(ConsoleColor.Black, Console.ForegroundColor)
            Console.WriteLine("An invocation after Assert")
        End Try
    End Sub

    <TestMethod>
    <ExpectedException(GetType(InvalidOperationException))> ' Compliant
    Public Sub AssertInFinallyWithCatch()
        Console.ForegroundColor = ConsoleColor.Red
        Try
            Throw New InvalidOperationException()
        Catch e As InvalidOperationException
            Console.WriteLine(Console.ForegroundColor)
        Finally
            Assert.AreEqual(ConsoleColor.Black, Console.ForegroundColor)
        End Try
    End Sub

    <TestMethod>
    <ExpectedException(GetType(InvalidOperationException))> ' Compliant
    Public Sub AssertInCatchWithFinally()
        Console.ForegroundColor = ConsoleColor.Red
        Try
            Throw New InvalidOperationException()
        Catch e As InvalidOperationException
            Assert.AreEqual(ConsoleColor.Black, Console.ForegroundColor)
        Finally
            Console.WriteLine(Console.ForegroundColor)
        End Try
    End Sub
End Class

Namespace Unrelated
    <AttributeUsage(AttributeTargets.Method)>
    Public Class ExpectedExceptionAttribute
        Inherits Attribute

        Public ReadOnly Property ExceptionType As Type

        Public Sub New(exceptionType As Type)
            Me.ExceptionType = exceptionType
        End Sub
    End Class
End Namespace

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ExpectedExceptionAttributeShouldNotBeUsed.MsTest.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Tests.Diagnostics
{
    class Program
    {
        [TestMethod]
        [ExpectedException(typeof(ArgumentNullException))]  // Noncompliant
        public void TestFoo1()
        {
            var x = true;
            x.ToString();
        }

        [TestMethod]
        [ExpectedExceptionAttribute(typeof(ArgumentNullException))]  // Noncompliant
        public void WithAttributeSuffix()
        {
            var x = true;
            x.ToString();
        }

        [TestMethod]
        [Microsoft.VisualStudio.TestTools.UnitTesting.ExpectedException(typeof(ArgumentNullException))]  // Noncompliant
        public void FullyQualifiedAttribute()
        {
            var x = true;
            x.ToString();
        }

        [TestMethod]
        [Unrelated.ExpectedException(typeof(ArgumentNullException))]  // Noncompliant - FP
        public void UnrelatedAttribute()
        {
            var x = true;
            x.ToString();
        }

        [TestMethod]
        [ExpectedException(typeof(ArgumentNullException))] // Compliant - one line
        public void TestFoo3()
        {
            new object().ToString();
        }

        [TestMethod]
        [ExpectedException(typeof(ArgumentNullException))] // FN
        public void TestMultilineFalseNegative()
        {
            {
                new object().ToString();
                new object().ToString();
                new object().ToString();
            }
        }

        [TestMethod]
        [ExpectedException(typeof(ArgumentNullException))]  // Compliant - one line
        public string TestFoo5() => new object().ToString();

        [TestMethod]
        public void TestFoo7()
        {
            bool callFailed = false;
            try
            {
                //...
            }
            catch (ArgumentNullException)
            {
                callFailed = true;
            }
        }

        [TestMethod]
        public void TestWithThrowsAssertation()
        {
            object o = new object();
            Assert.ThrowsException<ArgumentNullException>(() => o.ToString());
        }
    }

    // https://github.com/SonarSource/sonar-dotnet/issues/8300
    class Repro_8300
    {
        [TestMethod]
        [ExpectedException(typeof(InvalidOperationException))] // Compliant - using ExpectedException makes the test more readable
        public void AssertInFinally()
        {
            Console.ForegroundColor = ConsoleColor.Red;
            try
            {
                throw new InvalidOperationException();
            }
            finally
            {
                Assert.AreEqual(ConsoleColor.Black, Console.ForegroundColor);
            }
        }

        [TestMethod]
        [ExpectedException(typeof(InvalidOperationException))] // Noncompliant
        public void NoAssertInFinally()
        {
            Console.ForegroundColor = ConsoleColor.Red;
            try
            {
                throw new InvalidOperationException();
            }
            finally
            {
                Console.WriteLine("No Assert");
            }
        }

        [TestMethod]
        [ExpectedException(typeof(InvalidOperationException))] // Noncompliant
        public void NoAssertInCatch()
        {
            Console.ForegroundColor = ConsoleColor.Red;
            try
            {
                throw new InvalidOperationException();
            }
            catch (InvalidOperationException e)
            {
                Console.ForegroundColor = ConsoleColor.Black;
            }
        }

        [TestMethod]
        [ExpectedException(typeof(InvalidOperationException))] // Compliant
        public void AssertInCatch()
        {
            Console.ForegroundColor = ConsoleColor.Red;
            try
            {
                throw new InvalidOperationException();
            }
            catch (InvalidOperationException e)
            {
                Assert.AreEqual(ConsoleColor.Black, Console.ForegroundColor);
            }
        }

        [TestMethod]
        [ExpectedException(typeof(InvalidOperationException))] // Compliant
        public void AssertInAllCatch()
        {
            Console.ForegroundColor = ConsoleColor.Red;
            try
            {
                throw new InvalidOperationException();
            }
            catch
            {
                Assert.AreEqual(ConsoleColor.Black, Console.ForegroundColor);
            }
        }

        [TestMethod]
        [ExpectedException(typeof(InvalidOperationException))] // Compliant
        public void AssertInAllCatch_InvocationBeforeAssert()
        {
            Console.ForegroundColor = ConsoleColor.Red;
            try
            {
                throw new InvalidOperationException();
            }
            catch
            {
                Console.WriteLine("An invocation before Assert");
                Assert.AreEqual(ConsoleColor.Black, Console.ForegroundColor);
            }
        }

        [TestMethod]
        [ExpectedException(typeof(InvalidOperationException))] // Compliant
        public void AssertInAllCatch_InvocationAfterAssert()
        {
            Console.ForegroundColor = ConsoleColor.Red;
            try
            {
                throw new InvalidOperationException();
            }
            catch
            {
                Assert.AreEqual(ConsoleColor.Black, Console.ForegroundColor);
                Console.WriteLine("An invocation after Assert");
            }
        }

        [TestMethod]
        [ExpectedException(typeof(InvalidOperationException))] // Compliant
        public void AssertInFinallyWithCatch()
        {
            Console.ForegroundColor = ConsoleColor.Red;
            try
            {
                throw new InvalidOperationException();
            }
            catch (InvalidOperationException e)
            {
                Console.WriteLine(Console.ForegroundColor);
            }
            finally
            {
                Assert.AreEqual(ConsoleColor.Black, Console.ForegroundColor);
            }
        }

        [TestMethod]
        [ExpectedException(typeof(InvalidOperationException))] // Compliant
        public void AssertInCatchWithFinally()
        {
            Console.ForegroundColor = ConsoleColor.Red;
            try
            {
                throw new InvalidOperationException();
            }
            catch (InvalidOperationException e)
            {
                Assert.AreEqual(ConsoleColor.Black, Console.ForegroundColor);
            }
            finally
            {
                Console.WriteLine(Console.ForegroundColor);
            }
        }
    }
}

namespace Unrelated
{
    [AttributeUsage(AttributeTargets.Method)]
    public class ExpectedExceptionAttribute : Attribute
    {
        public Type ExceptionType { get; private set; }

        public ExpectedExceptionAttribute(Type exceptionType)
        {
            ExceptionType = exceptionType;
        }
    }
}
